### PR TITLE
Fairlinks and event display visualisation

### PR DESCRIPTION
**Pull request by wiechula**
Enable the usage of FairLinks for MC matching for hits and digits
Add director and first class for hit visualisation using the event display.

It can be used in the following way:
run simple simulation
`tpc-run-sim -m all`

start display in root
`
gSystem->AddIncludePath("-I$O2_ROOT/include -I$FAIRROOT_ROOT/include");
.x $PATH_TO_O2_SOURCE/tpcEventDisplay.C+("AliceO2_TGeant3.tpc.mc_2_event.root","AliceO2_TGeant3.tpc.params_2.root")
`

Click 'FairEventManager' -> 'Info' -> 'Current Event' arrow up

### CODE REVIEW

File: Detectors/TPC/eventdisplay/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+set(HEADERS
+   include/${MODULE_NAME}/HitDraw.h
+)
+Set(LINKDEF src/TPCEventdisplayLinkDef.h)
+Set(LIBRARY_NAME ${MODULE_NAME})
+set(BUCKET_NAME tpc_eventdisplay_bucket)
+
+O2_GENERATE_LIBRARY()
+
+#install(
```
Comment (sawenzel) [inline Detectors/TPC/eventdisplay/CMakeLists.txt]: can these lines be deleted?

Comment (wiechula) [inline Detectors/TPC/eventdisplay/CMakeLists.txt]: you mean the commented ones? sure. I'll update the pull request

File: Detectors/TPC/simulation/include/TPCSimulation/DigitCRU.h [hunk 2]
```diff
@@ -41,6 +41,10 @@ class DigitCRU{
     /// \return Size of the time bin container
     size_t getSize() const {return mTimeBins.size();}
 
+    /// Get the container
+    /// \return container
+    const std::deque<std::unique_ptr<DigitTime>>& getContainer() const { return mTimeBins; }
+
```
Comment (sawenzel) [inline Detectors/TPC/simulation/include/TPCSimulation/DigitCRU.h]: getContainer sounds a bit too generic. Can we consider having a more concrete name? (getTimeBins or getTimeBinContainer ...)

File: Detectors/TPC/simulation/include/TPCSimulation/DigitPad.h [hunk 3]
```diff
@@ -39,11 +42,15 @@ class DigitPad{
...
+    /// Get the MC Links
+    /// \return MC Links
+    const FairMultiLinkedData& getMCLinks() const { return mMCLinks; }
+
     /// Add digit to the time bin container
     /// \param eventID MC ID of the event
     /// \param trackID MC ID of the track
     /// \param charge Charge of the digit
-    void setDigit(int eventID, int trackID, float charge);
+    void setDigit(size_t hitID, float charge);
```
Comment (matthiasrichter) [inline Detectors/TPC/simulation/include/TPCSimulation/DigitPad.h]: adjust parameter description

File: Detectors/TPC/simulation/include/TPCSimulation/DigitPad.h [hunk 4]
```diff
@@ -53,41 +60,35 @@ class DigitPad{
...
-    /// This is then dumped into a std::vector and attached to the digits
-    /// \todo Find out how many different event/track combinations are relevant
-    /// \param std::vector containing the sorted MCLabels
-    void processMClabels(std::vector<long> &sortedMCLabels);
-    
   private:
     float                  mChargePad;   ///< Total accumulated charge on that pad for a given time bin
     unsigned char          mPad;         ///< Pad of the ADC value
-    std::map<long, int>    mMCID;        ///< Map containing the MC labels (key) and the according number of occurrence (value)
+    FairMultiLinkedData    mMCLinks;     ///< MC links
```
Comment (sawenzel) [inline Detectors/TPC/simulation/include/TPCSimulation/DigitPad.h]: Why do we need this data member: Isn't this provided by inheriting from FairMultiLinkedData_Interface?
The Interface already holds a pointer to the link data and provides functionality to access and retrieve the links.

File: Detectors/TPC/simulation/include/TPCSimulation/DigitTime.h [hunk 5]
```diff
@@ -55,7 +59,7 @@ class DigitTime{
     /// \param row Pad row of digit
     /// \param pad Pad of digit
     /// \param charge Charge of the digit
-    void setDigit(int eventID, int trackID, int cru, int row, int pad, float charge);
+    void setDigit(size_t hitID, int cru, int row, int pad, float charge);
```
Comment (matthiasrichter) [inline Detectors/TPC/simulation/include/TPCSimulation/DigitTime.h]: adjust parameter documentation

File: Detectors/TPC/simulation/include/TPCSimulation/DigitPad.h [hunk 6]
```diff
@@ -53,41 +60,35 @@ class DigitPad{
...
 inline 
-void DigitPad::setDigit(int eventID, int trackID, float charge)
+void DigitPad::setDigit(size_t hitID, float charge)
 {
   /// the MC ID is encoded such that we can have 999,999 tracks
   /// numbers larger than 1000000 correspond to the event ID
   /// i.e. 12000010 corresponds to event 12 with track ID 10
   /// \todo Faster would be a bit shift
-  ++mMCID[(eventID)*1000000 + trackID];
+  mMCLinks.AddLink(FairLink(-1, FairRootManager::Instance()->GetEntryNr(), "TPCPoint", hitID));
```
Comment (matthiasrichter) [inline Detectors/TPC/simulation/include/TPCSimulation/DigitPad.h]: is `"TPCPoints"` just a descriptive information? I suggest adding a constant for this

Comment (sawenzel) [inline Detectors/TPC/simulation/include/TPCSimulation/DigitPad.h]: It is the branch name in the ROOT file. I agree that we might change this later...

File: Detectors/TPC/eventdisplay/include/TPCEventdisplay/HitDraw.h [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+#include "Rtypes.h"                     // for HitDraw::Class, etc
+
+class TObject;
+class TVector3;
+
+namespace o2 {
+namespace TPC {
+
+class HitDraw: public FairPointSetDraw
```
Comment (sawenzel) [inline Detectors/TPC/eventdisplay/include/TPCEventdisplay/HitDraw.h]: not crucial for the moment: But in principle I would find the following name "TPCHitDrawTask" a better name.
Then it is clear what we are drawing + that this thing is actually a FairTask.

File: Detectors/TPC/simulation/include/TPCSimulation/DigitPad.h (hunk 4 above)
Comment (wiechula) [inline Detectors/TPC/simulation/include/TPCSimulation/DigitPad.h]: DigitPad is our own sorting container and does not inherit from from FairMultiLinkedData_Interface.
`class DigitPad{`
When we format the output data (DigitMC) we use exactly this inheritance:
`
void DigitPad::fillOutputContainer(TClonesArray *output, int cru, int timeBin, int row, int pad, float commonMode)
{
  /// The charge accumulated on that pad is converted into ADC counts, saturation of the SAMPA is applied and a Digit is created in written out
  const float totalADC = mChargePad - commonMode; // common mode is subtracted here in order to properly apply noise, pedestals and saturation of the SAMPA

  const float mADC = SAMPAProcessing::makeSignal(totalADC, PadSecPos(CRU(cru).sector(), PadPos(row, pad)));
  if(mADC > 0) {
    TClonesArray &clref = *output;
    const size_t digiPos = clref.GetEntriesFast();
    DigitMC *digit = new(clref[digiPos]) DigitMC(cru, mADC, row, pad, timeBin, commonMode);
    digit->SetLinks(getMCLinks());
  }
}
`

File: Detectors/TPC/simulation/include/TPCSimulation/DigitPad.h (hunk 6 above)
Comment (wiechula) [inline Detectors/TPC/simulation/include/TPCSimulation/DigitPad.h]: Yes, we discussed this during the meeting. I think this should be handled in the same way for all detectors. So for me this is framework.

File: Detectors/TPC/eventdisplay/include/TPCEventdisplay/HitDraw.h [hunk 8]
```diff
@@ -0,0 +1,43 @@
+/********************************************************************************
+ *    Copyright (C) 2014 GSI Helmholtzzentrum fuer Schwerionenforschung GmbH    *
```
Comment (wiechula) [inline Detectors/TPC/eventdisplay/include/TPCEventdisplay/HitDraw.h]: ok, I'll check

File: Detectors/TPC/simulation/include/TPCSimulation/DigitPad.h (hunk 4 above)
Comment (sawenzel) [inline Detectors/TPC/simulation/include/TPCSimulation/DigitPad.h]: Fair enough. Did not see this. That re-raises the question of the usefulness of the FairMultiLinkedData_Interface...

File: Detectors/TPC/eventdisplay/include/TPCEventdisplay/HitDraw.h [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+ ********************************************************************************/
+/*
+ * FairMCPointsDraw.h
+ *
+ *  Created on: Apr 17, 2009
+ *      Author: stockman
+ */
+
+#ifndef FAIRMCPOINTDRAW_H_
+#define FAIRMCPOINTDRAW_H_
```
Comment (sawenzel) [inline Detectors/TPC/eventdisplay/include/TPCEventdisplay/HitDraw.h]: these macro names are wrong (they will collide with macros from FairRoot).

File: Detectors/TPC/eventdisplay/include/TPCEventdisplay/HitDraw.h [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+class HitDraw: public FairPointSetDraw
+{
+  public:
+    HitDraw();
+    HitDraw(const char* name, Color_t color ,Style_t mstyle, Int_t iVerbose = 1):FairPointSetDraw(name, color, mstyle, iVerbose) {};
+    virtual ~HitDraw();
+
+  protected:
+    TVector3 GetVector(TObject* obj);
```
Comment (sawenzel) [inline Detectors/TPC/eventdisplay/include/TPCEventdisplay/HitDraw.h]: I believe this needs to be marked "override"

File: Detectors/TPC/eventdisplay/include/TPCEventdisplay/HitDraw.h [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+
+    ClassDef(HitDraw,0);
```
Comment (sawenzel) [inline Detectors/TPC/eventdisplay/include/TPCEventdisplay/HitDraw.h]: This needs to be made ClassDefOverride

**Comment by sawenzel** [comment]
@ktf: alibot is spamming this page

**Comment by sawenzel** [comment]
We are blocked merging this because the event display classes of FairRoot are only available when 
Geant3 is found. One of our requirements is to be able to build even when this is not the case.
So, we probably need to add a protection in our CMake.

On the other hand, I made an issue on the FairRoot side:

https://github.com/FairRootGroup/FairRoot/issues/550

and I believe they should hide this (spurious) dependency on their side and offer  the event display classes always.

**Comment by sawenzel** [comment]
@wiechula : It might be best to split this PR into links + display PR so that we can merge the links.

**Comment by ktf** [comment]
The open question is however how (and if) we make sure there is no dependency on the simulation required by reconstruction.

**Comment by wiechula** [comment]
@sawenzel : I can try to split this commit.
@ktf : I think I don't understand the comment

**Comment by MohammadAlTurany** [comment]
@ktf We used Geane (Track propagation ) to propagate reconstructed tracks (or Particles without visualisation info stored in the MC)

**Comment by wiechula** [comment]
Ok, for the moment I'll close this pull request and open a new one with only the FairLinks until it is clarified how to deal with the visualisation stuff

**Comment by sawenzel** [comment]
Concerning visualization: My understanding is anyways that we'll be using  separate event display developed inside O2 itself.

**Pull request closed without merging**